### ISSUE TITLE
Restructure shared libraries: lib/ tree with laygen + posgen

### ISSUE THREAD

**Issue by shunk031** [creative-graphic-design/design-generators#64]
Proposal (from the project owner) to restructure the shared-library layout before it ossifies:

## Idea

- Move shared libraries out of `models/` into a dedicated `lib/` tree (e.g. `lib/common/...`), keeping `models/` purely for model packages.
- Rename `layout_generation_common` to a shorter, brandable library name: **`laygen`**, used as a top-level namespace with a `common` submodule — `from laygen.common import LayoutGenerationOutput` — leaving room for future submodules like `laygen.datasets` / `laygen.testing` in the same namespace (e.g. `lib/laygen/src/laygen/common/`).
- Add a sibling **`posgen`** library (same shape: `posgen.common`) for the poster-generation common role (content-aware/poster shared pieces: content containers, saliency handling, CGL/PKU label specs, poster-specific visualization), so poster models (#14, #15, #44, #21, #22, #48, ...) share code the way layout models share `laygen`.

## Points to settle

1. Directory shape: `models/common` (status quo) vs `lib/laygen` + `lib/posgen` vs `lib/common/<name>` — and what `[tool.uv.workspace] members` globs become
2. Package naming: `layout-generation-common`/`layout_generation_common` vs `laygen`; import ergonomics; PyPI-style name collision check
3. Scope split: what belongs to `laygen` vs `posgen` vs individual model packages; dependency direction (`posgen` depends on `laygen`?)
4. Migration timing: PR #61 (which introduces `models/common`) is still open with five stacked PRs on top — renaming before merge is the cheapest point; decide fast
5. Backward compatibility: none needed pre-merge; if post-merge, alias policy

A design council (multiple agents, opposing stances) will discuss and post a recommendation below.

**Comment by shunk031** [comment]
# Structure recommendation

Chair: `structure-migration`. Outcome: recommend updating PR #61 before merge to move the shared code from `models/common` into a `lib/` library tree, rename the public package surface to `laygen.common`, and add the symmetric poster namespace `posgen.common`.

## 1. Final recommendation

Use this directory shape:

```text
lib/
  laygen/
    pyproject.toml
    src/laygen/
      __init__.py
      common/
        __init__.py
        bbox.py
        discrete.py
        labels.py
        outputs.py
        testing.py
        visualization.py
    tests/
      test_common.py
  posgen/
    pyproject.toml
    src/posgen/
      __init__.py
      common/
        __init__.py
        content.py
        labels.py
        testing.py
        visualization.py
    tests/
      test_common.py
models/
  layout-dm/
  const-layout/
  lace/
  layout-flow/
  layoutformerpp/
  layout-corrector/
```

Root `pyproject.toml`:

```toml
[project]
dependencies = [
    "laygen",
    "layout-dm",
    "transformers[torch,vision]>=5.0.0",
]

[tool.uv.workspace]
members = ["lib/*", "models/*"]

[tool.uv.sources]
laygen = { workspace = true }
design-generators-posgen = { workspace = true }
layout-dm = { workspace = true }
```

`lib/laygen/pyproject.toml`:

```toml
[project]
name = "laygen"
version = "0.1.0"
description = "Shared layout generation schemas and utilities."
requires-python = ">=3.11"
dependencies = [
    "matplotlib>=3.8",
    "numpy>=1.24",
    "torch>=2.1",
    "transformers>=5.0.0",
]
```

`laygen.common.outputs.LayoutGenerationOutput` must be based on
`transformers.utils.ModelOutput`, not `diffusers.utils.BaseOutput`, so the shared
library does not pull in a diffusers dependency:

```python
from dataclasses import dataclass
from typing import Any

import torch
from transformers.utils import ModelOutput


@dataclass
class LayoutGenerationOutput(ModelOutput):
    bbox: torch.Tensor
    labels: torch.Tensor
    mask: torch.Tensor
    id2label: dict[int, str]
    sequences: torch.Tensor | None = None
    scores: torch.Tensor | None = None
    trajectory: Any | None = None
    intermediates: Any | None = None
```

`lib/posgen/pyproject.toml`:

```toml
[project]
name = "design-generators-posgen"
version = "0.1.0"
description = "Shared poster generation schemas and utilities."
requires-python = ">=3.11"
dependencies = [
    "laygen",
    "numpy>=1.24",
    "torch>=2.1",
]

[tool.uv.sources]
laygen = { workspace = true }
```

Use `design-generators-posgen` as the distribution name if public package collision matters. The import package remains `posgen`.

Import examples:

```python
from laygen.common import LayoutGenerationOutput
from laygen.common import normalize_boxes, id2label_for_dataset
from laygen.common.bbox import ltwh_to_xywh, ltrb_to_xywh
from laygen.common.discrete import log_onehot_to_index, sample_categorical
from laygen.common.testing import assert_layout_output_schema

from posgen.common import PosterContent, normalize_content_image
from posgen.common.labels import poster_id2label_for_dataset
```

Existing PR #61 module mapping:

| Current module | Final module |
| --- | --- |
| `layout_generation_common.outputs` | `laygen.common.outputs` |
| `layout_generation_common.bbox` | `laygen.common.bbox` |
| `layout_generation_common.labels` | `laygen.common.labels` |
| `layout_generation_common.discrete` | `laygen.common.discrete` |
| `layout_generation_common.visualization` | `laygen.common.visualization` |
| `layout_generation_common.testing` | `laygen.common.testing` |

`laygen.common.__init__` should re-export the high-frequency public surface:

```python
from .bbox import clamp_boxes, denormalize_boxes, normalize_boxes
from .bbox import ltwh_to_xywh, ltrb_to_xywh, xywh_to_ltwh, xywh_to_ltrb
from .labels import id2label_for_dataset, label2id_for_dataset, labels_for_dataset
from .outputs import LayoutGenerationOutput
```

## 2. Scope allocation

| Concern | `laygen.common` | `posgen.common` | Model packages |
| --- | --- | --- | --- |
| Public output | `LayoutGenerationOutput` and schema rules from issue #2 | No separate output schema; may use/re-export `LayoutGenerationOutput` only if needed | Fill optional `sequences`, `scores`, `trajectory`, `intermediates` |
| Bbox | `xywh`/`ltwh`/`ltrb` conversion, normalization, denormalization, clamp, linear discretize/continuize | Calls `laygen.common.bbox`; no duplicate bbox math | Checkpoint-specific bins, coordinate token order, vendor parity logic |
| Labels | RICO25, PubLayNet, Magazine, Crello and generic dataset helpers | CGL/CGL-v2/PKU poster labels, aliases, CGL spelling canonicalization, PKU `INVALID` policy | Persist exact checkpoint `id2label`; handle private pad/background ids |
| Discrete helpers | log-onehot, categorical sampling, top-k/top-p, extraction helpers | None initially | Tokenizers and schedulers stay local |
| Content image | None | `PosterContent`, `normalize_content_image()` | Image resizing, feature extraction, saliency model outputs, checkpoint tensors |
| Retrieval | Defer unless it proves generic beyond posters | Defer until RALF/PosterO implementation proves exact common container | Retrieval table/cache lookup and model-specific feature tensors |
| Parsing | None | Defer JSON/HTML/SVG helpers until two LLM-style poster models share behavior | Prompt templates, generated text parsing, repair/retry loops |
| Visualization | Generic normalized layout renderer | Poster/background overlay wrapper if trivial | Debug plots and parity artifacts |
| Testing | Schema, bbox, mask, label, generator reproducibility assertions | Poster label/content assertions | Vendor parity, smoke tests, model/processor/tokenizer behavior |

Dependency direction:

```text
models/*       -> laygen
poster models  -> posgen -> laygen
laygen         -> no posgen or model imports
posgen         -> no model imports
```

## 3. Migration plan

Do the structural change in PR #61 before merge.

1. PR #61 owner moves `models/common` to `lib/laygen`.
2. PR #61 owner renames `src/layout_generation_common` to `src/laygen/common`.
3. PR #61 owner adds `src/laygen/__init__.py` and `src/laygen/common/__init__.py` with the re-exports above.
4. PR #61 owner changes `lib/laygen/pyproject.toml` project name from `layout-generation-common` to `laygen`.
5. PR #61 owner adds minimal `lib/posgen` with `src/posgen/common/{content,labels,testing,visualization}.py`.
6. PR #61 owner changes root workspace members to `["lib/*", "models/*"]`.
7. PR #61 owner replaces root source key `layout-generation-common` with `laygen`; adds `design-generators-posgen` if included in root dependencies.
8. PR #61 owner updates `models/layout-dm/pyproject.toml` to depend on `laygen`.
9. PR #61 owner rewrites LayoutDM imports from `layout_generation_common.*` to `laygen.common.*`.
10. PR #61 owner moves common tests to `lib/laygen/tests`, adds import smoke tests, and adds minimal `lib/posgen/tests`.
11. PR #61 owner runs `uv lock`, `uv run pre-commit run --all-files`, `uv run pytest`, `uv run --package laygen pytest`, `uv run --package design-generators-posgen pytest`, and `uv run --package layout-dm pytest`.
12. Stacked implementers rebase their branches on the updated #61 base, then do only mechanical import/dependency rewrites before continuing feature work.

Notification text for stacked implementers:

```text
Heads up: issue #64 changes the shared package before PR #61 merges.

Use:
  from laygen.common import LayoutGenerationOutput
  from laygen.common.bbox import ...
  from laygen.common.labels import ...
  from laygen.common.testing import ...

Do not import layout_generation_common or depend on layout-generation-common.
Workspace members are now ["lib/*", "models/*"].

Poster/content-aware helpers may import posgen.common, but keep model-specific processors, feature extraction, retrieval cache lookup, prompt parsing, tokenizers, schedulers, and model configs local unless a helper already exists in posgen.common.

If your branch is based on the old #61 shape, rebase after #61 updates and do a mechanical import/dependency rewrite before adding new shared abstractions.
```

Fallback if #61 merges before this change:

1. Open a dedicated migration PR from latest `origin/main`.
2. Add a temporary `layout_generation_common` compatibility shim only if already-merged code imports it.
3. Remove the shim after stacked branches rebase and merged packages use `laygen.common`.

## 4. Rejected alternatives

Rejected: keep `models/common` and `layout_generation_common` unchanged.

Reason: it conflicts with the user's clarified direction for a top-level `laygen.common` namespace and leaves shared libraries under the model package tree. The churn concern is real, but the cheapest way to handle it is a single pre-merge rewrite in #61.

Rejected: flat imports such as `laygen.outputs` and `posgen.content`.

Reason: the user clarified that `laygen` should be a top-level namespace with `laygen.common` now and room for future `laygen.datasets`, `laygen.testing`, etc. Poster side should mirror this as `posgen.common`.

Rejected: `lib/common/laygen` and `lib/common/posgen`.

Reason: the clarified Python namespace already expresses common-ness. `lib/laygen` and `lib/posgen` keep workspace globs simple with `["lib/*", "models/*"]` and avoid an extra filesystem level.

Rejected: large initial `posgen.common` with saliency, retrieval, parsing, and prompt schemas.

Reason: poster demand is real, but exact shared behavior is not proven yet. Start with labels/content/testing and add saliency/retrieval/parsing when #14/#44 or another poster PR needs concrete shared code.

Rejected: compatibility alias before #61 merges.

Reason: there is no released or merged public import yet. Adding both `layout_generation_common` and `laygen.common` before merge creates two names without a compatibility need.

## 5. Unresolved points

- Confirm packaging policy for `posgen`: use `design-generators-posgen` distribution name unless maintainers confirm public `posgen` ownership or no-publication policy.
- Decide whether root `[project].dependencies` should include `design-generators-posgen` immediately or only when the first poster package imports it. Workspace source can exist either way.
- Confirm whether `posgen.common.visualization` should be created in the initial PR or deferred with saliency/retrieval/parsing. It is safe only if implemented as a thin wrapper over `laygen.common.visualization`.
- Confirm exact implementation branch list for the five stacked model implementers. The migration plan assumes `const-layout`, `lace`, `layout-flow`, `layoutformerpp`, and `layout-corrector` are based on `impl-layout-dm`.

**Comment by shunk031** [comment]
Adopted (owner decision), with naming resolution: distribution/package names are plain **`laygen`** and **`posgen`** (not `design-generators-*`). Final shape: `lib/laygen/src/laygen/common/{outputs,outputs_diffusers,bbox,labels,discrete,visualization,testing}.py` and a thin `lib/posgen/src/posgen/common/` skeleton; workspace members `["lib/*", "models/*"]`; `posgen` is not added to root dependencies until its first consumer lands. The move happens pre-merge inside PR #61; stacked branches follow with a mechanical import rewrite (`layout_generation_common` → `laygen.common`).

**Comment by creative-graphic-design-dev** [comment]
The shared-library-structure canon now lives at https://github.com/creative-graphic-design/design-generators/blob/main/docs/architecture.md; this issue is preserved as history.
